### ISSUE TITLE
Sound not playing on iPhone (5/6) on first visit ever

### ISSUE THREAD

**Issue by markusf** [goldfire/howler.js#395]
Hey guys,

we currently have some issues with our app on first visit:

Whenever the website is opened the first time (really the first ever visit) on iPhone 5/6 we cannot play any sounds. When the page is reloaded it is suddenly working. I also checked that the iosautoenable Function is called successfully, it is.

Furthermore I also put in Charles (proxy) inbetween to enable no caching. But somehow it is also not related to that. As soon as I switch the domain, the sound is not playing the first time.

Did anybody experience similar issues?

Also tried with version 2, still not working.

**Comment by erlinghoff** [comment]
We are experiencing the same issue, on Safari for os X and on the IOS version. Sounds will never play on the first visit.

I don't know if this can have anything to do with the sound format fallback stack, as howler previously would not play anything in safari when the .opus versions of sound files were defined in the stack. Our current format fallback stack is:

.opus
.ogg
.m4a
.mp3

**Comment by markusf** [comment]
We are just using mp3 for now, no fallbacks defined. But good too know (unfortunately) somebody else having the same issues.

**Comment by erlinghoff** [comment]
Ok,  we can eliminate that the fallback is the cause of this, then.

**Comment by erlinghoff** [comment]
marks: have you tested your app on safari for os x? Just wondering if you will experience the same problem on os x safari, like we are.

**Comment by markusf** [comment]
No, we just tested on mobile, but I check later on Mac.

**Comment by goldfire** [comment]
Any update on this? Are you saying you can clear all data on the browser and visit again and it works fine, but if you change the domain it won't work? Basically, is there a way to test this without changing the domain on every single test?

**Comment by sokeroner** [comment]
Hi,
I have the same problem on Safari (9.0 (11601.1.56)) Mac OS X (10.11).
The problem occurs on the first use after the launch of safari.
If I close safari and then I reopen it, the problem re-occurs.

I use howler v1.1.28 on my app, but the problem is the same with [your website](http://goldfirestudios.com/blog/104/howler.js-Modern-Web-Audio-Javascript-Library)

**Comment by sokeroner** [comment]
The problem is the same with the v 2.0 sample.

After the launch of safari :
first load -> KO
second load -> OK

the problem is not global, so i think it's a cache problem

1. Launch of safari
2. first load of my app -> KO
3. second load of my app -> OK
4. first load of your website -> KO
5. second laod of your website -> OK

**Comment by sokeroner** [comment]
strange... if the web inspector is open, the audio is working well on the first use..

**Comment by markusf** [comment]
I only could reproduce by changing the domain, also with no-cache I could not reproduce.

**Comment by sokeroner** [comment]
I made some tests and this libraries/examples have the same problem:

- [webaudiox](https://github.com/jeromeetienne/webaudiox)
- [wad](https://github.com/rserota/wad#live-demo)
- [simple example](http://www.williammalone.com/posts/ios-9-web-audio-api-safari-mobile-issue/)

but this is working :

- [jsbin](http://jsbin.com/retibahuva/1/edit?html,css,js,output)
- [jsbin original example](http://jsbin.com/lenapigo/1/edit?html,css,js,output)

**Comment by sokeroner** [comment]
hum.. After investigation, it seems that tests work on jsbin only because rendering is done in an iframe that reloads content..

**Comment by sokeroner** [comment]
Maybe I found an answer... 

With the latest version of webkit, a user action is required to create the AudioContext. The problem is that Howler created the audiocontext when it is instantiated.

For me, this simple test example works with the lastest safari and webkit nightly on Mac OSX (10.11):

```javascript
var simpleWebAudioPlayer = function() {
    "use strict";

    var player = {},
        sounds = [],
        ctx,
        masterGain;

    player.load = function(sound) {
        sounds[sound.name] = sound;
        // Load the sound
        var request = new window.XMLHttpRequest();
        request.open("get", sound.src, true);
        request.responseType = "arraybuffer";
        request.onload = function() {
            ctx.decodeAudioData(request.response, function(buffer) {
                sounds[sound.name].buffer = buffer;
                if (sounds[sound.name].callback) {
                    sounds[sound.name].callback();
                }
            });
        };
        request.send();
    };

    player.play = function(name) {
        var inst = {};
        if (sounds[name]) {
            inst.source = ctx.createBufferSource();
            inst.source.buffer = sounds[name].buffer;
            inst.source.connect(masterGain);
            inst.source.start(0);
        }
    };

    // Create audio context
    if (typeof AudioContext !== 'undefined') {
        ctx = new AudioContext();
    } else if (typeof webkitAudioContext !== 'undefined') {
        ctx = new webkitAudioContext();
    }
    masterGain = (typeof ctx.createGain === 'undefined') ? ctx.createGainNode() : ctx.createGain();
    masterGain.gain.value = 1;
    masterGain.connect(ctx.destination);

    return player;
};


window.addEventListener('mousedown', createSound);

function createSound() {
    var player = new simpleWebAudioPlayer();
    player.load({
        name: 'son1',
        src: 'audio/test.mp3',
        callback: function() {
            player.play('son1');
        }
    });
}
```

If I do not use the mousedown event, this example has the same problem as Howler...

**Comment by Jip-Hop** [comment]
I had a similar problem but for me it was on every page load. After the first user interaction, the sound would play muted. After the second user interaction I could hear audio.

I worked around this issue by playing a silent mp3 audio file at page load and calling .unload(). This way my sounds will play on the first user interaction.

Maybe the workaround applies to your issue too?

```
var sound = new Howl({
    src: 'data:audio/mp3;base64,SUQzAwAAAAAAFgAAAAAAAAAAAAAAAAAAAAAAAAAAAAD/4xjAAAAAAlwAAAAAtASxAAAACAAATQAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAD/4xjAMQAAAlwAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAD/4xjAbAAAAlwAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA='
});
sound.unload();

$('body').click(function(){
    sound = new Howl({
      src: ['sound.mp3']
    });
    sound.play();
});
```
I now continue to use `sound` for any sounds I want to play. It seems that this way the AudioContext is re-created upon user interaction and becomes un-muted.

**Comment by goldfire** [comment]
Based on my testing, this should now be fixed with 93e09dc102f9ad507339d69e3d48e00869e2e373. What I found was that the AudioContext was created, but it defaulted to suspended. Simply calling `resume` on the context fixed this, which happens by default now when playing (however, since `Howler`'s state defaulted to `running`, this wasn't called previously).

**Comment by EvgeniyRRU** [comment]
> Based on my testing, this should now be fixed with 93e09dc.
Dear @goldfire do you know, is there the same fix in the 1.1 branch? Or I should to adapt my codebase to 2.0 version? Nasty bug in Safari, isn't?

**Comment by ayamflow** [comment]
Apparently it's still not working, just had the same experience, using the npm v2 beta version.

On desktop I've been using [this fix](http://stackoverflow.com/a/34823090) with success.

**Comment by goldfire** [comment]
@ayamflow Do you have a test case where this doesn't work?

**Comment by ayamflow** [comment]
@goldfire Actually nevermind, I just realized the npm version doesn't update itself, we were stuck on `2.0.0-beta1`. Sorry!

**Comment by kriss145** [comment]
Hi,

I have similiar problem. When I want to autoplay audio on the mobile that is the problem.
I need to tap on the screen (everywhen) to play audio..

Could help?

**Comment by themoonrat** [comment]
There is nothing you can do about this. Mobile browsers **must** have a user interaction before audio is allowed to play.

**Comment by inear** [comment]
And soon also desktop when Canary grows up into M66...

**Comment by DavidZernik** [comment]
TLDR - Is there a way to fix this? I use howler.js for my web application and the audio doesn't work on iPhone web application. Should I just replace howler.js entirely??

**Comment by st-h** [comment]
@DavidZernik just read through this issue. Seems like this has been fixed already and other causes and fixes have been mentioned.

If you experience a similar issue with current version and have gone through this issue, I recommend opening a new one and provide sufficient details like your code that is causing issues, as well as howler and browser versions.

If you experience that you have different use cases than what howler has been designed for, replacing howler might be a good idea. However, you should probably  expect encountering bugs in browser implementations, that you might need to work around.
